src/sentry/reprocessing2.py

```diff
@@ -167,26 +169,68 @@ def is_reprocessed_event(data):
 def _get_original_event_id(data):
     from sentry.event_manager import get_tag
 
+    # XXX: Get rid of this tag once we reuse event IDs
     return get_tag(data, "original_event_id")
 
 
-def should_save_reprocessed_event(data):
-    if not data:
-        return False
+def _get_original_group_id(data):
+    from sentry.event_manager import get_tag
+
+    # XXX: Have real snuba column
+    return get_tag(data, "original_group_id")
+
 
-    orig_id = _get_original_event_id(data)
+def _get_sync_redis_client():
+    return redis_clusters.get(settings.SENTRY_REPROCESSING_SYNC_REDIS_CLUSTER)
 
-    if not orig_id:
-        return True
 
-    orig_event = eventstore.get_event_by_id(project_id=data["project"], event_id=orig_id)
+def _get_sync_counter_key(group_id):
+    return "re2:count:{}".format(group_id)
+
+
+def mark_event_reprocessed(data):
+    """
+    This function is supposed to be unconditionally called when an event has
+    finished reprocessing, regardless of whether it has been saved or not.
+    """
+    if not is_reprocessed_event(data):
+        return
 
-    if not orig_event:
-        return True
+    key = _get_sync_counter_key(_get_original_group_id(data))
+    _get_sync_redis_client().decr(key)
 
-    for prop in ("threads", "stacktrace", "exception", "debug_meta"):
-        old_data = orig_event.data.get(prop)
-        if old_data != data.get(prop):
-            return True
 
-    return False
+def start_group_reprocessing(project_id, group_id, max_events=None):
+    from sentry.models.group import Group, GroupStatus
+    from sentry.models.grouphash import GroupHash
+    from django.db import transaction
+
+    with transaction.atomic():
+        Group.objects.filter(id=group_id).update(status=GroupStatus.REPROCESSING)
+        # Remove all grouphashes such that new events get sorted into a
+        # different group.
+        GroupHash.objects.filter(group_id=group_id).delete()
+
+    # Get event counts of issue (for all environments etc). This was copypasted
+    # and simplified from groupserializer.
+    event_count = snuba.aliased_query(
+        aggregations=[["count()", "", "times_seen"]],  # select
+        dataset=snuba.Dataset.Events,  # from
+        conditions=[["group_id", "=", group_id], ["project_id", "=", project_id]],  # where
+        referrer="reprocessing2.start_group_reprocessing",
+    )["data"][0]["times_seen"]
```

Comment:
Yeah that's why I used `if count <= 0` instead of `if count == 0` when waiting for reprocessing to be done. The idea was that, because this query is not actually enforcing an upper bound on timestamps or anything, there's a high chance those events will also end up being reprocessed properly before group deletion runs, because the redis countdown for waiting with group deletion is polled only every 5 minutes. So you have two race conditions cancelling each other out.

I just realized that doesn't actually work if the snuba query returns a few events and doesn't have to paginate. A solution would be to sleep 1 minute between deleting grouphashes and starting the query.